File: mydirectory/ansario_hw06/buell_hw06 2/Message.java
import java.util.TreeMap;
/*********************************************************************
 * Message class for Homework 06.
 *
 * This class has instance variables for message-centric data as
 * well as an <code>ArrayList</code> for the packets inside a message.
 *
 * Copyright(C) 2013 Duncan A. Buell.  All rights reserved.
 *
 * @author Duncan Buell
 * @version 1.00 2013-02-17 
 *
**/
public class Message<T extends Comparable<T>> implements IMessage
{
  private final int DUMMYINT = -999;
  private int messageID;
  private int packetCount;
  private TreeMap<Integer, Packet> packetMap;

/*********************************************************************
 * Constructor
**/
  public Message()
  {
    this.messageID = DUMMYINT;
    this.packetCount = DUMMYINT;
    packetMap = new TreeMap<Integer, Packet>();
  } // public Message()

/*********************************************************************
 * Accessors and Mutators.
**/

/*********************************************************************
 * Accessor for <code>messageID</code>.
 *
 * @return the message ID
**/
  public int getMessageID()
  {
    return this.messageID;
  } // public int getMessageID()

/*********************************************************************
 * General methods
**/

/*********************************************************************
 * Method to test if the message contains a given packet.
 *
 * @param p the packet to test
 * @return the answer to the question.
**/
  public boolean contains(Packet p)
  {
    Packet packetToFind = this.packetMap.get(p.getPacketID());

    if(null != packetToFind)
      return true;
    else
      return false;
  } // public boolean contains(Packet p)

/*********************************************************************
 * Method to insert a packet into this message.
 *
 * @param p the packet to insert into this message
**/
  public void insert(Packet p)
  {
    this.packetMap.put(p.getPacketID(), p);
    this.messageID = p.getMessageID();
    this.packetCount = p.getPacketCount();
  } // public void insert(Packet p)

/*********************************************************************
 * Method to test if the message is complete.
 *
 * @return the answer to the question.
**/
  public boolean isComplete()
  {
    if(this.packetCount == this.packetMap.size())
      return true;
    else
      return false;
  } // public boolean isComplete()

/*********************************************************************
 * Usual <code>toString</code> method.
 *
 * @param messageID the id of the message to be written.
 * @return a 'String' version of the message
**/
  public String toString()
  {
    String s = "";

    s += String.format("ID: %4d%n", this.messageID);
    s += String.format("PacketCount: %4d%n", this.packetCount);
    s += String.format("Nodes:      %d%n", this.packetMap.size());

    for(Integer packetID : this.packetMap.keySet())
    {
      s += String.format("%s%n", this.packetMap.get(packetID));
    }

    return s;
  } // public String toString()

} // public class Message implements IMessage

File: testdirectory/ansario_hw06/Message.cpp
#include "Message.h"

/******************************************************************************
 *3456789 123456789 123456789 123456789 123456789 123456789 123456789 123456789
 * Class 'Message' for a single message.
 *
 *This class represents a message which will be sent by being broken down into  
 * packets. It is the storage container which will be used to separate the dif-
 * ferent messages sent at different times.
 *
 * Author: Duncan A. Buell Edited by Omar A. Ansari
 * Date last modified: 4 November 2014
**/

/******************************************************************************
 * Constructor
**/
Message::Message()
{
	messageID = 0;
	packetCount = 0;
}

/******************************************************************************
 * Destructor
**/
Message::~Message()
{
}

/******************************************************************************
 * Accessors and Mutators
**/

/******************************************************************************
 * Accessor 'getMessageID'.
 *
 * Returns:
 *   the 'messageID'
**/
int Message::getMessageID() const
{
	return messageID;
}

/******************************************************************************
 * General functions.
**/
/******************************************************************************
 * Method to test if a message contains a given packet.
 *
 * Parameters:
 *   p - the 'Packet' to be tested
 *
 * Returns:
 *   true if the message containes the packet parameter. False if not.
**/
bool Message::contains(const Packet& p) const
{
	if (message.count(p.getPacketID()) == 1)
	{
		return true;
	} else 
	{
		return false;
	}
}

/******************************************************************************
 * Method to insert a packet into a message. This will update the packet count 
 * and messageID.
 * Parameters:
 *   p - the 'Packet' to be tested
 *
 * Returns:
 *   nothing.
**/
void Message::insert(const Packet& p)
{
	message.insert(pair<int, Packet>(p.getPacketID(), p));
	this->packetCount=p.getPacketCount();
	this->messageID = p.getMessageID();
	//message.packetCount = p.getPacketCount();
}

/******************************************************************************
 * Method to test whether a message is complete. This will test if the packetCount
 * is equal to the overall size of the message.
 * Returns:
 *   true if size of message == packetCount. False if not.
**/
bool Message::isComplete() const
{
	if (message.size()-1 == this->packetCount)
	{
		return true;
	} else
	{
		return false;
	}
}

/******************************************************************************
 * Function 'toString'.
 *
 * Returns:
 *   the 'toString' of the message as a 'vector' of packets
**/
string Message::toString() const
{
	string s = "";
	s += ("MessageID: " + getMessageID());
	//s += ("Nodes: " + message.size() + "\n");

	map<int, Packet>::const_iterator iter;

	for (iter = this->message.begin(); iter != this->message.end(); ++iter)
	{
		s += Utils::Format((*iter).second.toString());
	}

	return s;
}
